cocotb/drivers/amba.py

```diff
@@ -60,153 +89,401 @@ def __init__(self, entity, name, clock, **kwargs):
+        #  * Transaction IDs to 0
+        #  * Region identifier to 0
+        #  * Normal (non-exclusive) access
+        #  * Device non-bufferable access
+        #  * Unprivileged, secure data access
+        #  * No QoS
+        unsupported_signals = [
+            "AWID", "AWREGION", "AWLOCK", "AWCACHE", "AWPROT", "AWQOS",
+            "ARID", "ARREGION", "ARLOCK", "ARCACHE", "ARPROT", "ARQOS"]
+        for signal in unsupported_signals:
+            try:
+                getattr(self.bus, signal).setimmediatevalue(0)
+            except AttributeError:
+                pass
+
+        # Mutex for each channel to prevent contention
+        self.write_address_busy = Lock(name + "_awbusy")
+        self.read_address_busy = Lock(name + "_arbusy")
+        self.write_data_busy = Lock(name + "_wbusy")
+        self.read_data_busy = Lock(name + "_rbusy")
+        self.write_response_busy = Lock(name + "_bbusy")
+
+    @staticmethod
+    def _check_length(length: int, burst: AXIBurst) -> None:
+        """Check that the provided burst length is valid."""
+        if length <= 0:
+            raise ValueError("Burst length must be a positive integer")
+
+        if burst is AXIBurst.INCR:
+            if length > 256:
+                raise ValueError("Maximum burst length for INCR bursts is 256")
+        elif burst is AXIBurst.WRAP:
+            if length not in (1, 2, 4, 8, 16):
+                raise ValueError("Burst length for WRAP bursts must be 1, 2, "
+                                 "4, 8 or 16")
+        else:
+            if length > 16:
+                raise ValueError("Maximum burst length for FIXED bursts is 16")
+
+    @staticmethod
+    def _check_size(size: int, data_bus_width: int) -> None:
+        """Check that the provided transfer size is valid."""
+        if size > data_bus_width:
+            raise ValueError("Beat size ({} B) is larger than the bus width "
+                             "({} B)".format(size, data_bus_width))
+        elif size <= 0 or size & (size - 1) != 0:
+            raise ValueError("Beat size must be a positive power of 2")
+
+    @staticmethod
+    def _check_4kB_boundary_crossing(address: int, burst: AXIBurst, size: int,
+                                     length: int) -> None:
+        """Check that the provided burst does not cross a 4kB boundary."""
+        if burst is AXIBurst.INCR:
+            last_address = address + size * (length - 1)
+            if address & ~0xfff != last_address & ~0xfff:
+                raise ValueError(
+                    "INCR burst with start address {:#x} and last address "
+                    "{:#x} crosses the 4kB boundary {:#x}, which is forbidden "
+                    "in a single burst"
+                    .format(address, last_address,
+                            (address & ~0xfff) + 0x1000))
+
+    async def _send_write_address(
+        self, address: int, length: int, burst: AXIBurst, size: int,
+        delay: int, sync: bool
```

Comment:
I understand what `sync` is *doing*, but I'm not sure what the point is. When would you not want to sync?

Reply:
No idea, `sync` was already present in the AXI4-Lite driver so I kept it. Maybe if you are already on a clock rising edge from a previous operation and you don't want to wait another one.